sdk/search/azure-search-index/design.md

```diff
@@ -0,0 +1,483 @@
+
+    def SearchIndexClient.batch_update(
+        self,
+        batch: IndexOperationBatch,
+        **kwargs
+    ) -> List[IndexingResult]:
+```
+
+### Query Building
+
+In addition to simple text searches, Azure Search supports more sophisticated
+search requests. These include ODATA queries for filtering, selecting, and
+ordering result and Lucene queries for search.
+
+#### Lucene Search Queries
+
+The search field of the service accepts Lucene query strings, which may of two
+formats: simple or full.
+
+#### Simple Lucene Query Syntax
+
+Propose that simple mode search queries, e.g. ``"wifi AND parking NOT luxury"``
+simply be supplied as-is by users. In this context an API is more likely to
+get in the way, rather than provide useful structure for users to rely on.
+
+#### Full Lucene Query Syntax
+
+TBD
+
+#### ODATA fields
+
+The API below is intended to allow users to build up complicated ODATA string
+expressions without having to construct the ODATA strings manually. These are
+building block components to be used to pass expressions to a ``SearchQuery``
+object. Examples after the API list demonstrate the functionality, as a
+reference.
+
+```python
+
+# Low level operations
+
+def NOT(x) -> str:
+def AND(*args) -> str:
+def OR(*args) -> str:
+def EQ(x, y) -> str:
+def NEQ(x, y) -> str:
+def GT(x, y) -> str:
+def GE(x, y) -> str:
+def LT(x, y) -> str:
+def LE(x, y) -> str:
+
+def ALL(path: str, expression: Union[str, callable]) -> str:
+def ANY(path: str, expression: Union[str, callable, None] = None) -> str:
+
+# supports comparisons
+class VAR(object):
+
+# ODATA operators
+
+def asc(sorter) -> str:
+def desc(sorter) -> str:
+
+def search_score() -> str
+
+class orderby:
+
+class geo_distance:
+class POINT:
+
+class geo_intersection:
+class POLYGON:
+
+class search_in:
+class search_ismatch:
+class search_ismatchscoring:
+
+class select:
+```
+
+##### Examples of ODATA string generation
+
+The focus of the API above may be better illustrated with examples. Each of the
+the expressions below can be passed to configure a ``SearchQuery``, which can
+then use them to resolve to the correct ODATA strings.
+
+```python
+>>> ob = orderby("Rating desc", "BaseRate", asc("stars"), asc(search_score()))
+>>> print(ob)
+$orderby=Rating desc,BaseRate,stars asc,search.score() asc
+
+>>> geo_distance("FOO", POINT(10, 20)
+>>> print(gd)
+geo.distance(FOO, geography'POINT(10 20)')
+
+>>> print(gd >= 10)
+geo.distance(FOO, geography'POINT(10 20)') ge 10
+
+>>> s = select("HotelId", "HotelName", "Rating", "Address/City")
+>>> print(s)
+$select=HotelId,HotelName,Rating,Address/City
+
+>>> si = search_in("FOO", "a", "b", "c", delimiter="|")
+>>> print(si)
+search.in(FOO, 'a|b|c', '|')
+
+>>> AND("Address/City eq 'Vancouver'",  "Address/Country eq 'US'", geo_distance("FOO", POINT(10, 20)) < 10)
+"(Address/City eq 'Vancouver' and Address/Country eq 'US' and geo.distance(FOO, geography'POINT(10 20)') lt 10)"
+
+>>> ALL("tags", lambda t: t == 'wifi')
```

Comment:
How complicated can the lambda be here? And what is the experience when I step out of what we can serialize into a query?

Reply:
@johanste I don't follow the second question. 

For the first, it should be able to handle arbitrary comparisons, and subsitutions into whatever subset of ODATA operations that are defined, e.g. here is another:
```
>>> ALL("locations", lambda loc: AND(loc == 'FOO', loc > 4, NOT(geo_distance(loc, POINT(10, 20)) > 10)))
"locations/all(loc: (loc eq 'FOO' and loc gt 4 and not (geo.distance(loc, geography'POINT(10 20)') gt 10)))"

```

Reply:
I'm just wondering what happens if I try to use something (completely nonsensical, for illustration purposes only) like `lambda t: hash(t) == '7'` in the filter expression. Based on anecdotal data, similar issues appeared in the past when it wasn't clear to the caller what could be serialized and executed on the service vs. things that couldn't. 

I assume that you would just get a `ValueError` with a very clear message in that case, is that correct?

Reply:
Well, presently you get a `TypeError` but there's a clear place to catch that and raise something more informative